external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala

```diff
@@ -18,222 +18,247 @@
-   */
-  def release(): Unit
+  override def close(): Unit = {
+    consumer.close()
+  }
 
-  /** Reference to the internal implementation that this wrapper delegates to */
-  protected def internalConsumer: InternalKafkaConsumer
-}
+  /** Create a KafkaConsumer to fetch records for `topicPartition` */
+  private def createConsumer: KafkaConsumer[Array[Byte], Array[Byte]] = {
+    val c = new KafkaConsumer[Array[Byte], Array[Byte]](kafkaParams)
+    val tps = new ju.ArrayList[TopicPartition]()
+    tps.add(topicPartition)
+    c.assign(tps)
+    c
+  }
 
+  private def seek(offset: Long): Unit = {
+    logDebug(s"Seeking to $groupId $topicPartition $offset")
+    consumer.seek(topicPartition, offset)
+  }
+}
 
 /**
- * A wrapper around Kafka's KafkaConsumer that throws error when data loss is detected.
- * This is not for direct use outside this file.
+ * The internal object to store the fetched data from Kafka consumer and the next offset to poll.
+ *
+ * @param _records the pre-fetched Kafka records.
+ * @param _nextOffsetInFetchedData the next offset in `records`. We use this to verify if we
+ *                                 should check if the pre-fetched data is still valid.
+ * @param _offsetAfterPoll the Kafka offset after calling `poll`. We will use this offset to
+ *                           poll when `records` is drained.
  */
-private[kafka010] case class InternalKafkaConsumer(
-    topicPartition: TopicPartition,
-    kafkaParams: ju.Map[String, Object]) extends Logging {
-  import InternalKafkaConsumer._
-
-  /**
-   * The internal object to store the fetched data from Kafka consumer and the next offset to poll.
-   *
-   * @param _records the pre-fetched Kafka records.
-   * @param _nextOffsetInFetchedData the next offset in `records`. We use this to verify if we
-   *                                 should check if the pre-fetched data is still valid.
-   * @param _offsetAfterPoll the Kafka offset after calling `poll`. We will use this offset to
-   *                           poll when `records` is drained.
-   */
-  private case class FetchedData(
-      private var _records: ju.ListIterator[ConsumerRecord[Array[Byte], Array[Byte]]],
-      private var _nextOffsetInFetchedData: Long,
-      private var _offsetAfterPoll: Long) {
-
-    def withNewPoll(
-        records: ju.ListIterator[ConsumerRecord[Array[Byte], Array[Byte]]],
-        offsetAfterPoll: Long): FetchedData = {
-      this._records = records
-      this._nextOffsetInFetchedData = UNKNOWN_OFFSET
-      this._offsetAfterPoll = offsetAfterPoll
-      this
-    }
-
-    /** Whether there are more elements */
-    def hasNext: Boolean = _records.hasNext
-
-    /** Move `records` forward and return the next record. */
-    def next(): ConsumerRecord[Array[Byte], Array[Byte]] = {
-      val record = _records.next()
-      _nextOffsetInFetchedData = record.offset + 1
-      record
-    }
+private[kafka010] case class FetchedData(
```

Comment:
I know it's old code but but this blocks to solve the multiple consumer extra seek problem.
If this would be readonly it doesn't matter which task gets which consumer. Consumer just takes a look at the offset indexed data pool for data, gives it back and makes prefetch again.

Reply:
Allowing fields to be mutable is the optimization for avoiding new instance for every batch. FetchedDataPool ensures each FetchedData doesn't be provided to multiple tasks concurrently, and _nextOffsetInFetchedData is updated once it is returning to pool so that it can be offset information for searching.

Please let me know if I'm missing something here.

Reply:
Now I see how the seeking issue solved, good. I think it's highly welcome and essential in terms of performance which would be good to cover at least in one test (made a toy test and in logs I see it works). Something like:
```
createTopicWithOnePartition
get(0)
assert seek
get(500)
assert seek
get(1)
assert no seek
```

Reply:
I guess we already have `multiple tasks referring same key continuously using fetched data` if we just want to verify correctness, though checking with KafkaDataConsumer may worth it in terms of acceptance test. `seek` itself doesn't impact huge on performance. `Discarding fetched data due to offset mismatch` would hugely impact on performance, so this should be checked if we would want to be sure about improvement.

Reply:
Just added some tests on KafkaDataConsumer, basic sanity as well as edge cases.
Please take a look at 4e2790a2cb6eff1b78af0f5a01946dad29b0b2ad

Reply:
Yeah, I think the tests are fine with minor suggestions. Thanks!